### PR TITLE
Costs

### PR DESCRIPTION
**Pull request by julian-evers**
Calculate emissions, material and production costs

### CODE REVIEW

File: pybamm_tea/tea.py [hunk 1]
```diff
@@ -303,14 +307,47 @@ def initialize(self):
...
+            raise ValueError("Error: 'Initial concentration in negative electrode [mol.m-3]' missing. Please, supply both or none.")
+        if (
+            pava.get("Initial concentration in negative electrode [mol.m-3]") is not None
+            and pava.get("Initial concentration in positive electrode [mol.m-3]") is None
+        ):
+            raise ValueError("Error: 'Initial concentration in positive electrode [mol.m-3]' missing. Please, supply both or none.")
+        if pava.get("Initial loss of lithium inventory") is None:
+            pava["Initial loss of lithium inventory"] = 0
+            warnings.warn("Warning: 'Initial loss of lithium inventory' is set to 0.")
+        if (
```
Comment (rtimms) [inline pybamm_tea/tea.py]: by this point neither should be None, based on the above logic

File: pybamm_tea/tea.py [hunk 2]
```diff
@@ -303,14 +307,47 @@ def initialize(self):
...
             )
-        self.parameter_values = {**self.parameter_values, **pava}
+        if (
+            pava.get("Initial concentration in negative electrode [mol.m-3]") is None
+            and pava.get("Initial concentration in positive electrode [mol.m-3]") is None
+        ):
+            # initialize concentrations based on initial loss of lithium inventory
+            pava["Initial concentration in negative electrode [mol.m-3]"] = 0
+            pava["Initial concentration in positive electrode [mol.m-3]"] = pava.get("Maximum concentration in positive electrode [mol.m-3]")
+            warnings.warn("Warning: 'Initial concentration in negative electrode [mol.m-3]' and 'Initial concentration in positive electrode [mol.m-3]' are set to 0 and maximum concentration in positive electrode, respectively.")
```
Comment (rtimms) [inline pybamm_tea/tea.py]: In practice the initial concentrations won't be 0 and c_max because you will charge/discharge between voltage limits. I think what you are really doing here is saying that the cyclable lithium is equivalent to starting with a fully lithiated cathode and fully delithiated anode. You could then do something like 
```python3
    param = pybamm.LithiumIonParameters()
    Q_n = parameter_values.evaluate(param.n.Q_init)
    Q_p = parameter_values.evaluate(param.p.Q_init)
    Q_Li = Q_p * f # here 0 < f < 1 gives you initial LLI
    inputs = {"Q_n": Q_n, "Q_p": Q_p, "Q_Li": Q_Li}
    esoh_solver = pybamm.lithium_ion.ElectrodeSOHSolver(parameter_values, param)
    sol = esoh_solver.solve(inputs)

    c_n_max = parameter_values.evaluate(param.n.prim.c_max)
    c_p_max = parameter_values.evaluate(param.p.prim.c_max)
    x = sol["x_100"]
    y = sol["y_100"]
    parameter_values.update(
        {
            "Initial concentration in negative electrode [mol.m-3]": x * c_n_max,
            "Initial concentration in positive electrode [mol.m-3]": y * c_p_max,
        },
        check_already_exists=False,
    )
```

File: pybamm_tea/tea.py [hunk 3]
```diff
@@ -303,14 +307,47 @@ def initialize(self):
...
+                    pava["Lithium inventory [mA.h.cm-2]"] = pava.get("Initial lithium inventory [mA.h.cm-2]") - float(pava.get("Initial loss of lithium inventory [mA.h.cm-2]"))
+                    pava["Initial concentration in negative electrode [mol.m-3]"] = pava.get("Initial concentration in negative electrode [mol.m-3]") * (1 - pava.get("Initial loss of lithium inventory"))
+                    pava["Initial concentration in positive electrode [mol.m-3]"] = pava.get("Initial concentration in positive electrode [mol.m-3]") * (1 - pava.get("Initial loss of lithium inventory"))
+                    pava["Initial stoichiometry"] = pava.get("Lithium inventory [mA.h.cm-2]") / ((pava.get("Negative electrode thickness [m]") * pava.get("Negative electrode active material volume fraction") * pava.get("Maximum concentration in negative electrode [mol.m-3]") + pava.get("Positive electrode thickness [m]") * pava.get("Positive electrode active material volume fraction") * pava.get("Maximum concentration in positive electrode [mol.m-3]")) * 96485 / 3.6 / 10000)
```
Comment (rtimms) [inline pybamm_tea/tea.py]: Is this what people normally mean be "initial stoichiometry"?

File: pybamm_tea/tea.py [hunk 4]
```diff
@@ -394,223 +425,191 @@ def print_capacities_and_potentials(self):
...
+            and pava.get("Negative electrode stoichiometry at 100% SoC") is None
+            and pava.get("Positive electrode stoichiometry at 100% SoC") is not None
+            and pava.get("Positive electrode stoichiometry at 0% SoC") is not None
         ):
             y100 = pava.get("Positive electrode stoichiometry at 100% SoC")
             y0 = pava.get("Positive electrode stoichiometry at 0% SoC")
-            # x0, x100 = self.get_stoichiometries(pava, y0, y100)
-            # update cut-off voltages if voltage curve(s) are provided
+            # calculate lithium in positive electrode at SOC = 0
+            n_pe_0 = y0 * (pava.get("Maximum concentration in positive electrode [mol.m-3]")
```
Comment (rtimms) [inline pybamm_tea/tea.py]: Would be cleaner to have already stored the electrode capacities, then you can just get those from the dict instead of repeating the calculation all the time

File: pybamm_tea/tea.py [hunk 5]
```diff
@@ -394,223 +425,191 @@ def print_capacities_and_potentials(self):
...
+                         * 96485 / 3.6 / 10000))
+            # calculate lithium in negative electrode at SOC = 1
+            x100 = ((pava.get("Lithium inventory [mA.h.cm-2]") - n_pe_100)
+                      / (pava.get("Maximum concentration in negative electrode [mol.m-3]")
+                         * pava.get("Negative electrode thickness [m]")
+                         * pava.get("Negative electrode active material volume fraction")
+                         * 96485 / 3.6 / 10000))
+            if x0 < 0 or x100 < 0 or x0 > 1 or x100 > 1:
+                raise ValueError("Error: Stoichiometry calculation for negative electrode failed.")
+            if pava.get("Negative electrode OCP [V]") is not None and pava.get("Positive electrode OCP [V]") is not None:
```
Comment (rtimms) [inline pybamm_tea/tea.py]: These voltage cut-offs are normally defined first for the cell, e.g. 2.5-4.2V, and then the stoichiometry windows are calculated to respect those voltage limits. This way seems backwards?

File: pybamm_tea/tea.py [hunk 6]
```diff
@@ -394,223 +425,191 @@ def print_capacities_and_potentials(self):
...
+    def calculate_stack_energy(self):
         """
         Calculate ideal volumetric and gravimetric energy densities on stack level.
         """
         stack_ed = {}  # stack energy densities dict
-        pava = None
-        pava = self.parameter_values  # parameter values
+        pava = dict(self.parameter_values)  # parameter values
 
         # stoichiometries - calculation based on input stoichiometries or cell
```
Comment (rtimms) [inline pybamm_tea/tea.py]: there is a lot of repeated code here and it is hard to follow. can you explain the use case for being able to independently change sto limits?

File: pybamm_tea/tea.py [hunk 7]
```diff
@@ -394,223 +425,191 @@ def print_capacities_and_potentials(self):
...
-            )  # stoichiometries at 0 and 100% SOC based on cell potential limits
+            )
         else:
             raise ValueError("Error: Stoichiometry calculation failed.")
         stack_ed["Negative electrode stoichiometry at 0% SoC"] = x0
         stack_ed["Negative electrode stoichiometry at 100% SoC"] = x100
         stack_ed["Positive electrode stoichiometry at 100% SoC"] = y100
         stack_ed["Positive electrode stoichiometry at 0% SoC"] = y0
 
+        # update initial concentrations in electrodes to SoC = 1
```
Comment (rtimms) [inline pybamm_tea/tea.py]: is this consistent wit previous calculations?

it seems like there are multiple places where the same values get calculated, and it is confusing

**Review by rtimms — CHANGES_REQUESTED** [review]
looking good, just a few comments. also, why do you use `get` everywhere instead of just indexing into the dict? you only really need to use `get` if you think the key might not be there and you want to provide a default when it isn't

File: pybamm_tea/tea.py (hunk 2 above)
Comment (julian-evers) [reply pybamm_tea/tea.py]: The idea was that _if_ no initial concentrations are supplied, the total lithium inventory is what would be supplied with the maximum concentration in the positive electrode before formation, without considering any losses at this stage. Later the initial concentrations are updated to 100% SoC, so that the TEA parameter-set can directly be used for a discharge simulation. Would you prefer to have the initial lithium inventory set according to 100% SoC (dependent on voltage cut-offs/stoichiometry limits)?

File: pybamm_tea/tea.py (hunk 1 above)
Comment (julian-evers) [reply pybamm_tea/tea.py]: Thank you for noticing, I'll remove the condition.

File: pybamm_tea/tea.py (hunk 3 above)
Comment (julian-evers) [reply pybamm_tea/tea.py]: I have no clue, but I don't think so, it is an artefact I delete now. (I don't remember exactly how I wanted to use that value.)

File: pybamm_tea/tea.py (hunk 6 above)
Comment (julian-evers) [reply pybamm_tea/tea.py]: It is useful for capacity balancing in case no OCV curves are supplied. I also generally like it to calculate voltage cut-offs based on it, one could also set target capacities based on it. I'll try to shorten the code section and add more comments so that it becomes easier to follow.
